src/KOKKOS/pair_snap_kokkos_impl.h

```diff
@@ -632,201 +698,145 @@ void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPComputeNeigh,const typen
-  SNAKokkos<DeviceType> my_sna = snaKK;
-
-  Kokkos::View<double*,Kokkos::LayoutRight,DeviceType,Kokkos::MemoryTraits<Kokkos::Unmanaged>>
-    d_coeffi(d_coeffelem,ielem,Kokkos::ALL);
-
-  for (int icoeff = 0; icoeff < ncoeff; icoeff++) {
-    d_beta_pack(iatom_mod,icoeff,iatom_div) = d_coeffi[icoeff+1];
-  }
+  const int ninside = d_ninside(ii);
+  if (jnbor >= ninside) return;
 
-  if (quadraticflag) {
-    const auto idxb_max = my_sna.idxb_max;
-    int k = ncoeff+1;
-    for (int icoeff = 0; icoeff < ncoeff; icoeff++) {
-      const auto idxb = icoeff % idxb_max;
-      const auto idx_chem = icoeff / idxb_max;
-      double bveci = my_sna.blist(idxb, idx_chem, ii);
-      d_beta_pack(iatom_mod,icoeff,iatom_div) += d_coeffi[k]*bveci;
-      k++;
-      for (int jcoeff = icoeff+1; jcoeff < ncoeff; jcoeff++) {
-        const auto jdxb = jcoeff % idxb_max;
-        const auto jdx_chem = jcoeff / idxb_max;
-        double bvecj = my_sna.blist(jdxb, jdx_chem, ii);
-        d_beta_pack(iatom_mod,icoeff,iatom_div) += d_coeffi[k]*bvecj;
-        d_beta_pack(iatom_mod,jcoeff,iatom_div) += d_coeffi[k]*bveci;
-        k++;
-      }
-    }
-  }
+  my_sna.compute_cayley_klein(iatom_mod,jnbor,iatom_div);
 }
 
-template<class DeviceType>
+template<class DeviceType, typename real, int vector_length>
 KOKKOS_INLINE_FUNCTION
-void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPPreUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPPreUi>::member_type& team) const {
-  SNAKokkos<DeviceType> my_sna = snaKK;
+void PairSNAPKokkos<DeviceType, real, vector_length>::operator() (TagPairSNAPPreUi, const int iatom_mod, const int j, const int iatom_div) const {
+  SNAKokkos<DeviceType, real, vector_length> my_sna = snaKK;
 
-  // Extract the atom number
-  const int ii = team.team_rank() + team.team_size() * (team.league_rank() % ((chunk_size+team.team_size()-1)/team.team_size()));
+  const int ii = iatom_mod + iatom_div * vector_length;
   if (ii >= chunk_size) return;
+
   int itype = type(ii);
   int ielem = d_map[itype];
 
-  my_sna.pre_ui(team,ii,ielem);
+  my_sna.pre_ui(iatom_mod, j, ielem, iatom_div);
 }
 
-template<class DeviceType>
+template<class DeviceType, typename real, int vector_length>
 KOKKOS_INLINE_FUNCTION
-void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPComputeUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPComputeUi>::member_type& team) const {
-  SNAKokkos<DeviceType> my_sna = snaKK;
+void PairSNAPKokkos<DeviceType, real, vector_length>::operator() (TagPairSNAPComputeUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPComputeUi>::member_type& team) const {
+  SNAKokkos<DeviceType, real, vector_length> my_sna = snaKK;
 
-  // Extract the atom number
-  int ii = team.team_rank() + team.team_size() * (team.league_rank() % ((chunk_size+team.team_size()-1)/team.team_size()));
-  if (ii >= chunk_size) return;
+  constexpr int team_size = sizeof(real) == 4 ? 8 : 4;
+  if (team.team_size() != team_size) return; // error
```

Comment:
If this check is true, does that mean there is a bug, and it should never get here? Or should error out?

Reply:
Ah, this is somewhat sloppy, it was meant to enforce consistency between that line and this line: https://github.com/weinbe2/lammps/blob/fea-snap-ui-aosoa/src/KOKKOS/pair_snap_kokkos_impl.h#L327

I can replace this with a `static constexpr int` that gets explicitly reused both places.

Reply:
Done, and amusingly removing that check gave a small speed-up.